src/pool.jl

```diff
@@ -81,10 +72,73 @@ function actual_free(buf::Mem.DeviceBuffer;
 end
 
 
-## memory limits
+## memory accounting
+
+struct MemoryStats
+  # maximum size of the memory heap
+  size::Int
+
+  # the amount of live bytes
+  live::Threads.Atomic{Int}
+
+  last_time::Threads.Atomic{Float64}
+  last_gc_time::Threads.Atomic{Float64}
+  last_freed::Threads.Atomic{Int}
+end
+
+function account!(stats::MemoryStats, bytes::Integer)
+  Threads.atomic_add!(stats.live, bytes)
+  if bytes > 0
+    Threads.atomic_add!(stats.live, 1)
+  end
+end
+
+const _memory_stats = PerDevice{MemoryStats}()
+function memory_stats(dev::CuDevice=device())
+  get!(_memory_stats, dev) do
+    device!(dev) do
+      MemoryStats(total_memory(), Threads.Atomic{Int}(0), Threads.Atomic{Float64}(0.0),
+                  Threads.Atomic{Float64}(0.0), Threads.Atomic{Int}(0),)
+    end
+  end
+end
+
+function maybe_collect()
+  stats = memory_stats()
 
-# to enforce a hard memory limit, we need to keep track of the amount currently allocated.
-const memory_use = Threads.Atomic{Int}(0)
+  # check that we're under memory pressure
+  pressure = stats.live[] / stats.size
+  if pressure < 0.75
+    return
+  end
+
+  # we tolerate 5% GC time. if we freed a lot last time, bump that up to 10%
+  last_time = stats.last_time[]
+  current_time = time()
+  gc_rate = stats.last_gc_time[] / (current_time - last_time)
+  allowed_gc_rate = if stats.last_freed[] > (0.1*stats.size)
+    0.1
+  else
+    0.05
+  end
+  if gc_rate > allowed_gc_rate
+    return
+  end
+  stats.last_time[] = current_time
+
+  # finally, call the GC
+  pre_gc_live = stats.live[]
+  gc_time = @elapsed GC.gc()
+  post_gc_live = stats.live[]
+  Threads.atomic_set!(stats.last_freed, memory_freed)
+  ## GC times can vary, so smooth them out
+  Threads.atomic_set!(stats.last_gc_time, 0.75*stats.last_gc_time[] + 0.25*gc_time)
```

Comment:
This is me going "now what would be a good GC heuristic" fully ignoring decades of research on the topic. So @gbaraldi if you have any thoughts; feel free.

Reply:
@DhairyaLGandhi reported that this didn't help his memory-hungry application, so the heuristics probably need tuning.